smart_home/src/services/temperature_service.cc

```diff
@@ -0,0 +1,66 @@
+#include <iostream>
+#include "../../include/services/temperature_service.h"
+#include "../../include/sensors/temperature_sensor.h"
+#include "../../include/devices/ac_unit.h"
+
+namespace services {
+
+void TemperatureService::AddSensor(sensors::Sensor* sensor) {
+    sensors_.push_back(sensor);
+}
+
+void TemperatureService::AddDevice(devices::Device* device) {
+    devices_.push_back(device);
+}
+
+void TemperatureService::Refresh() {
+    for (sensors::Sensor* sensor : sensors_) {
+        std::cout << "Refreshing..." << std::endl;
+        sensors::TemperatureSensor* current_sensor = dynamic_cast<sensors::TemperatureSensor*>(sensor);
```

Comment:
Be careful when using dynamic_cast, you're basically loading a shotgun and pointing it to your face.
This code right here tells me there's a design issue in your code:
Either the Sensor interface is poorly designed or you're trying to do a workaround for a bad decision along the way.
From a quick read from the code, it's the first. Even though sensors is an abastract class/interface, the classes that implement it don't have a uniform way of working. We'll cover this a bit more in the next lesson.

Reply:
On your last point that the temperature service has design flaws, a better approach would be to keep a list of devices for every sensor. For each device added, check which room does it correspond to and and add it to that sensor's list. Or another idea would be to have a vector of rooms, each room containing a sensor and a list of temperature related devices.